phpstorm/phpstorm.md

```diff
@@ -2,10 +2,13 @@
 title: PhpStorm Plugin Development
 ---
 
-Plugins for PhpStorm are developed in Java using IntelliJ IDEA (Community Edition is sufficient in this case). You will also need a copy of PhpStorm to develop against. The [PsiViewer plugin](https://plugins.jetbrains.com/plugin/227-psiviewer) will also be useful.
+Plugins for PhpStorm are developed in Java using either IntelliJ IDEA or PhpStorm. IntelliJ IDEA Community Edition does not support the PHP plugin.
+IntelliJ IDEA Community Edition can be used to develop a PhpStorm plugin, but IntelliJ IDEA Ultimate must be used for testing.
+ 
+You will also need a copy of PhpStorm to develop against. The [PsiViewer plugin](https://plugins.jetbrains.com/plugin/227-psiviewer) will also be useful.
```

Comment:
Let's clarify what we mean by _targeting a product_ vs. _which product can be used for development_ vs. _how to define the SDK classpaths_.  Here is my understanding:
* **Target:** These pages are about developing a plugin that targets PhpStorm. Discussion about a PHP plugin targeting (extending) IntelliJ IDEA seems to be confusing matters, so for now let's remove it.
*  **IDE:** PhpStorm plugins are written in Java, so IntelliJ IDEA is required. 
   * What are the PHP development limitations in IJ CE? For example, will IJ CE build a PhpStorm plugin project against an SDK that includes PHP libraries in the class path? 
   * When you say testing must be done in IntelliJ IDEA Ultimate rather than CE, does that mean CE does not support debugging a PhpStorm plugin?
   * If testing should be done in PhpStorm anyway, then let's explicitly say the product is also required.
*  **SDK:** If PhpStorm is required for testing, then why wouldn't a PhpStorm plugin project SDK always be based on the installed version of PhpStorm? (Rather than the installed version of IntelliJ.)

If you would please clarify these points then I will update phpstorm.md and setting_up_environment.md.  TIA

Reply:
We probably should come up with a consistent set of terms for the two IDEs that come into play when developing a plugin: the IDE in which you write the plugin code ("development IDE") and the IDE which you run to test the plugin ("target IDE").

As the development IDE, you can use IntelliJ IDEA Community Edition; there are no limitations.

As the target IDE, you can't use IntelliJ IDEA Community Edition, because it won't load the PHP plugin, and therefore your plugin which depends on PHP won't work either. You can use IntelliJ Ultimate with the PHP plugin, but it's better if you use PhpStorm, because in that way you avoid the risk of using some APIs which aren't available in PhpStorm.

If you're using the devkit project set up, you can create an IntelliJ Platform plugin SDK based on the installation of any target IDE, including PhpStorm, and indeed you should create it based on an installation of PhpStorm.

Reply:
Here's my understanding of developing PHP and PhpStorm plugins...
Development IDE: the IDE in which you write and build the plugin code.
Test IDE:  the IDE in which you are able to install and run the plugin for testing.
Target IDE: the IDE in which a _user_ will run the plugin.

JetBrains Product | Development IDE | Test IDE | Target IDE
------------ | ------------- | ------------- | -------------
IntelliJ CE | Y | N (1) | N (1)
IntelliJ UE | Y (2) | Y (2) | Y (2,3)
PhpStorm | N (4) | Y  | Y (5)

1. IntelliJ CE doesn't support the PHP plugin
2. Requires PHP plugin. 
3. Best practice is to base the plugin project SDK on an IntelliJ installation.
4. PhpStorm does not support Java development.
5. Best practice is to base the plugin project SDK on PhpStorm installation.

How do developers of PhpStorm plugins typically debug their code? Is debugging done within a development instance of IntelliJ?